backtracking.go

```diff
@@ -32,13 +32,13 @@ type Backtracking struct {
 	initG    float64
 }
 
-func (b *Backtracking) Init(loc LinesearchLocation, step float64) EvaluationType {
+func (b *Backtracking) Init(f, g float64, step float64) EvaluationType {
+	if g >= 0 {
+		panic("backtracking: initial derivative is non-negative")
+	}
```

Comment:
Could you move this after the step check? I can see the case where someone got both signs wrong, so the step is negative and the gradient is positive. Seeing the step check first is more useful.